### PR TITLE
Evaluate Api examples

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes are proposed in this pull request?
Evaluate Api examples

## How is this patch tested?

N/A

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: examples/evaluators/evaluate_on_regressor.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from mlflow.models.evaluation import evaluate, EvaluationDataset
+import mlflow
+from sklearn.datasets import load_boston
+from sklearn.linear_model import LinearRegression
+
+mlflow.sklearn.autolog()
+
+X, y = load_boston(return_X_y=True)
+dataset = EvaluationDataset(data=X, labels=y, name='boston', feature_names=[
+    f'feature_{i}' for i in range(13)
```
Comment (BenWilson2) [inline examples/evaluators/evaluate_on_regressor.py:10]: for feature_names, do we want to show the actual names from the source X df for the example?
```python
boston_data = load_boston()
dataset = EvaluationDataset(data=boston_data.data, labels=boston_data.target, name='boston', feature_names=boston_data.feature_names)
```

File: mlflow/models/evaluation/base.py [hunk 2]
```diff
@@ -219,6 +229,24 @@ def __init__(self, data, labels, name=None, path=None):
...
+                if num_features != len(feature_names):
+                    raise ValueError('feature name list must be the same length with feature data.')
+                self._feature_names = feature_names
+            else:
+                self._feature_names = [f'f_{i}' for i in range(num_features)]
+        else:
+            pd_column_names = [c for c in self.data.columns if c != self.labels]
+            if feature_names is not None:
+                feature_names = list(feature_names)
+                if pd_column_names != list(feature_names):
```
Comment (BenWilson2) [inline mlflow/models/evaluation/base.py:245]: do we want to do a set comparison here to prevent list traversal for large feature sets?

Comment (BenWilson2) [inline mlflow/models/evaluation/base.py:245]: do we want to do a set comparison here to prevent list traversal for large feature sets?

File: mlflow/models/evaluation/base.py [hunk 3]
```diff
@@ -456,11 +485,120 @@ def _start_run_or_reuse_active_run(run_id):
...
+
+    def check_nesting_config_dict(_evaluator_name_list, _evaluator_name_to_conf_map):
+        return isinstance(_evaluator_name_to_conf_map, dict) and \
+               all(k in _evaluator_name_list and isinstance(v, dict)
+                   for k, v in _evaluator_name_to_conf_map.items())
+
+    if evaluators is None:
+        evaluator_name_list = list(_model_evaluation_registry._registry.keys())
+        if len(evaluator_name_list) > 1:
+            print(f'Multiple registered evaluators are found {evaluator_name_list} and '
```
Comment (BenWilson2) [inline mlflow/models/evaluation/base.py:507]: Do we want to use warning.warn() instead of stdout print so that it's easier for users to hide this message or send it to logs for automated frameworks that interface with this?

Comment (BenWilson2) [inline mlflow/models/evaluation/base.py:507]: Do we want to use warning.warn() instead of stdout print so that it's easier for users to hide this message or send it to logs for automated frameworks that interface with this?

File: mlflow/models/evaluation/base.py [hunk 4]
```diff
@@ -496,33 +634,17 @@ def evaluate(
                              a nested dictionary whose key is the evaluator name.
     :return: An :py:class:`mlflow.models.evaluation.EvaluationDataset` instance containing
              evaluation results.
+
+    The default evaluator supports the 'regressor' and 'classifer' `model_type`s. The available
+    `evaluator_config` options for the default evaluator include:
+     - log_model_explainability: The number of rows to use for calculating model explainability.
```
Comment (BenWilson2) [inline mlflow/models/evaluation/base.py:640]: I think this docstring is wrong. Isn't `explainability_nsamples` the target number of rows to use? Shouldn't this be a boolean flag for whether to log the results?

Comment (BenWilson2) [inline mlflow/models/evaluation/base.py:640]: I think this docstring is wrong. Isn't `explainability_nsamples` the target number of rows to use? Shouldn't this be a boolean flag for whether to log the results?

File: mlflow/models/evaluation/base.py [hunk 5]
```diff
@@ -496,33 +634,17 @@ def evaluate(
...
+     - explainality_algorithm: A string to specify the shap explainer algorithm. If not set, it will
```
Comment (BenWilson2) [inline mlflow/models/evaluation/base.py:641]: Do we want to explain what the mapping is or the location in the code that this default mapping is defined at?

File: mlflow/models/evaluation/base.py [hunk 6]
```diff
@@ -229,14 +257,7 @@ def data(self):
...
-
-            spark_df_type = SparkDataFrame
-        except ImportError:
-            spark_df_type = None
-
-        if spark_df_type and isinstance(self._original_data, spark_df_type):
+        if self._spark_df_type and isinstance(self._original_data, self._spark_df_type):
             self._data = self._original_data.limit(
                 EvaluationDataset.SPARK_DATAFRAME_LIMIT
             ).toPandas()
```
Comment (BenWilson2) [inline mlflow/models/evaluation/base.py:263]: Have we verified how this behaves if the feature to evaluate is actually contained in the output from a VectorAssembler transformer stage? If the feature set is a single-column SparseVector or DenseVector, translating back to columns with a provided set of field names might be a useful quality of life feature. 
If this is completely out of scope, disregard this comment.

Comment (BenWilson2) [inline mlflow/models/evaluation/base.py:263]: Have we verified how this behaves if the feature to evaluate is actually contained in the output from a VectorAssembler transformer stage? If the feature set is a single-column SparseVector or DenseVector, translating back to columns with a provided set of field names might be a useful quality of life feature. 
If this is completely out of scope, disregard this comment.

File: mlflow/models/evaluation/default_evaluator.py [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+                self._evaluate_per_class(None, self.y, self.y_pred, self.y_prob)
+                self._log_image_artifact(
+                    lambda: plot_lift_curve(self.y, self.y_probs), "lift_curve",
+                )
+            else:
+                self.metrics['f1_score_micro'] = \
+                    sk_metrics.f1_score(self.y, self.y_pred, average='micro')
+                self.metrics['f1_score_macro'] = \
+                    sk_metrics.f1_score(self.y, self.y_pred, average='macro')
+                for postive_class in range(self.num_classes):
```
Comment (BenWilson2) [inline mlflow/models/evaluation/default_evaluator.py:395]: awesome!

Comment (BenWilson2) [inline mlflow/models/evaluation/default_evaluator.py:395]: awesome!

File: mlflow/models/evaluation/default_evaluator.py [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+    def _evaluate_regressor(self):
+        self.y_pred = self.model.predict(self.X)
+        self.metrics["example_count"] = len(self.X)
+        self.metrics["mean_absolute_error"] = sk_metrics.mean_absolute_error(self.y, self.y_pred)
+        self.metrics["mean_squared_error"] = sk_metrics.mean_squared_error(self.y, self.y_pred)
+        self.metrics["root_mean_squared_error"] = math.sqrt(self.metrics["mean_squared_error"])
+        self.metrics['sum_on_label'] = sum(self.y)
+        self.metrics['mean_on_label'] = self.metrics['sum_on_label'] / self.metrics["example_count"]
+        self.metrics['r2_score'] = sk_metrics.r2_score(self.y, self.y_pred)
+        self.metrics['max_error'] = sk_metrics.max_error(self.y, self.y_pred)
```
Comment (BenWilson2) [inline mlflow/models/evaluation/default_evaluator.py:427]: we MAY want to guard on this depending on how the tracking server handles floating point errors. max_error can be + / - inf

File: mlflow/models/evaluation/base.py (hunk 2 above)
Comment (WeichenXu123) [reply mlflow/models/evaluation/base.py:245]: feature name should be unique, and we should also compare the feature order. so i think we should compare list

File: mlflow/models/evaluation/base.py (hunk 5 above)
Comment (WeichenXu123) [reply mlflow/models/evaluation/base.py:641]: The default algorithm behavir depends on shap Explainer implementation. I refine the doc like:
```
If not set, it will run `shap.Explainer` with "auto" algorithm, which choose the best fit explainer according to the model.
```

File: mlflow/models/evaluation/base.py (hunk 6 above)
Comment (WeichenXu123) [reply mlflow/models/evaluation/base.py:263]: Good point. 
We cannot support array-like feature value now. The Vector column in spark dataframe will be converted into a `Vector` instance in python side.

@dbczumar Shall we support a feature column which value is type of "ndarray" or list or "pyspark.ml.linalg.Vector" ?

File: mlflow/models/evaluation/default_evaluator.py (hunk 8 above)
Comment (WeichenXu123) [reply mlflow/models/evaluation/default_evaluator.py:427]: I just tried logging some special value:
```
with mlflow.start_run() as run:
  mlflow.log_metric('inf', np.inf)
  mlflow.log_metric('neg_inf', -np.inf)
  mlflow.log_metric('nan', np.nan)
```
mlflow can display them, shall we do additional handling ?
image

File: mlflow/models/evaluation/base.py (hunk 6 above)
Comment (dbczumar) [reply mlflow/models/evaluation/base.py:263]: @WeichenXu123 Is there any additional challenge to supporting `ndarray` or `list`? For `pyspark.ml.linalg.Vector`, is the issue that `df.toPandas()` doesn't convert the vector to a list / ndarray?

Comment (BenWilson2) [reply mlflow/models/evaluation/base.py:263]: For information purposes:
```python
from pyspark.ml.feature import VectorAssembler

data = spark.table("ben_demo.wine_data")
assembler = VectorAssembler(inputCols=[col for col in data.columns if col != "class"], outputCol="features")
data_assembled = assembler.transform(data)
collected = data_assembled.toPandas()



type(collected['features'][0])

>> pyspark.ml.linalg.DenseVector
```
We'll have to do a bit of conversion here...
```python

data_convert_vector = data_assembled.withColumn("feature_array", vector_to_array("features"))
data_columns = [col for col in data_convert_vector.columns if col != "features"]
data_to_collect = data_convert_vector.select(*data_columns)
collected_data = data_to_collect.toPandas()


type(collected_data['feature_array'][0])

>> numpy.ndarray
```

Another tidbit: if we don't do the explicit casting or leave a DenseVector or SparseVector in the dataframe, Arrow will throw a warning that it can't process the vectorized representation and will default back to the much slower older version of collection of a Spark DataFrame to a Pandas DataFrame.

File: mlflow/models/evaluation/default_evaluator.py (hunk 8 above)
Comment (BenWilson2) [reply mlflow/models/evaluation/default_evaluator.py:427]: haha good to know that it can handle those just fine! No concerns from me :)

File: mlflow/models/evaluation/base.py (hunk 5 above)
Comment (BenWilson2) [reply mlflow/models/evaluation/base.py:641]: Ah yes the internal deciding logic within SHAP. sounds good

File: mlflow/models/evaluation/base.py (hunk 6 above)
Comment (dbczumar) [reply mlflow/models/evaluation/base.py:263]: Got it. For Spark, it sounds like we can document this shortcoming for now and let it fail. For Pandas, I'm still not sure I see the challenge with supporting `ndarray` and `list` columns.

File: mlflow/models/evaluation/default_evaluator.py (hunk 8 above)
Comment (dbczumar) [reply mlflow/models/evaluation/default_evaluator.py:427]: MLflow is built to handle infs and nans. I don't think we need to do anything special here.

File: mlflow/models/evaluation/base.py (hunk 3 above)
Comment (dbczumar) [reply mlflow/models/evaluation/base.py:507]: @WeichenXu123 Can we use `_logger.warn()` / `_logger.info()` throughout this PR, instead of `print()`? We shouldn't use print in MLflow.

Comment (dbczumar) [reply mlflow/models/evaluation/base.py:507]: @WeichenXu123 Can we use `_logger.warn()` / `_logger.info()` throughout this PR, instead of `print()`? We shouldn't use print in MLflow.

File: mlflow/models/evaluation/lift_curve.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
+import matplotlib.pyplot as plt
+
+import numpy as np
+
+
+def _cumulative_gain_curve(y_true, y_score, pos_label=None):
+    """
+    This method is copied from scikit-plot package.
```
Comment (harupy) [inline mlflow/models/evaluation/lift_curve.py:8]: Can we add a permalink?

File: mlflow/models/evaluation/base.py (hunk 6 above)
Comment (WeichenXu123) [reply mlflow/models/evaluation/base.py]: ndarray/list/pyspark.ml.Vector as element type all supported now.

Comment (WeichenXu123) [reply mlflow/models/evaluation/base.py]: All kinds of columns includes "pyspark.ml.linalg.Vector" type are supported.

File: examples/evaluators/evaluate_on_binary_classifier.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import xgboost
```
Comment (dbczumar) [inline examples/evaluators/evaluate_on_binary_classifier.py:1]: @WeichenXu123 Can we change the directory name to `examples/evaluation/`?

**Review by dbczumar — APPROVED** [review]
@WeichenXu123 Can we change the directory name to examples/evaluation/? Can we also add a README file to this folder with a short explanation of the feature and an explanation of what each of the example Python scripts does? 

LGTM after that!

File: examples/evaluation/README.md [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+Run from the current git directory with Python.
+**Note**:  These examples assumes that you have all the dependencies for `scikit-learn`, `xgboost`, and `shap` library
+installed in your development environment.
+
+ `python evaluate_on_binary_classifier.py`
+ 
+ `python evaluate_on_multiclass_classifier.py`
+ 
+ `python evaluate_on_regressor.py`
```
Comment (harupy) [inline examples/evaluation/README.md:23]:
````suggestion

```
python evaluate_on_binary_classifier.py`
python evaluate_on_multiclass_classifier.py`
python evaluate_on_regressor.py`
````

File: examples/evaluation/README.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+specified dataset using builtin default evaluator, and log resulting metrics & artifacts to MLflow Tracking.
+
+- Example `evaluate_on_binary_classifier.py` evaluate a xgboost `XGBClassifier` model on dataset loaded by
+  `shap.datasets.adult`.
+- Example `evaluate_on_multiclass_classifier.py` evaluate a scikit-learn `LogisticRegression` model on dataset
+  generated by `sklearn.datasets.make_classification`.
+- Example `evaluate_on_regressor.py` evaluate a scikit-learn `LinearRegression` model on dataset loaded by
+  `sklearn.datasets.load_boston`
+
+#### How to run this code
```
Comment (harupy) [inline examples/evaluation/README.md:13]:
```suggestion
#### Prerequisites

```
pip install scikit-learn xgboost shap
```

#### How to run this code
```

Do we need matplotlib or not?

File: examples/evaluation/README.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+Run from the current git directory with Python.
+**Note**:  These examples assumes that you have all the dependencies for `scikit-learn`, `xgboost`, and `shap` library
+installed in your development environment.
```
Comment (harupy) [inline examples/evaluation/README.md:17]:
````suggestion
#### Prerequisites

```
pip install scikit-learn xgboost shap
```

#### How to run the scripts

````

File: examples/evaluation/README.md [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+The three simple examples illustrates how you can use the `mlflow.evaluate` API to evaluate a PyFunc model on the
+specified dataset using builtin default evaluator, and log resulting metrics & artifacts to MLflow Tracking.
+
+- Example `evaluate_on_binary_classifier.py` evaluate a xgboost `XGBClassifier` model on dataset loaded by
+  `shap.datasets.adult`.
+- Example `evaluate_on_multiclass_classifier.py` evaluate a scikit-learn `LogisticRegression` model on dataset
+  generated by `sklearn.datasets.make_classification`.
+- Example `evaluate_on_regressor.py` evaluate a scikit-learn `LinearRegression` model on dataset loaded by
+  `sklearn.datasets.load_boston`
```
Comment (harupy) [inline examples/evaluation/README.md:11]:
```suggestion
- `evaluate_on_binary_classifier.py` evaluates an xgboost `XGBClassifier` model on dataset loaded by
  `shap.datasets.adult`.
- `evaluate_on_multiclass_classifier.py` evaluates a scikit-learn `LogisticRegression` model on dataset.
  generated by `sklearn.datasets.make_classification`.
- `evaluate_on_regressor.py` evaluates a scikit-learn `LinearRegression` model on dataset loaded by
  `sklearn.datasets.load_boston`
```

File: examples/evaluation/evaluate_on_regressor.py [hunk 15]
```diff
@@ -0,0 +1,30 @@
+import mlflow
+from sklearn.datasets import load_boston
+from sklearn.linear_model import LinearRegression
+from sklearn.model_selection import train_test_split
+
+mlflow.sklearn.autolog()
+
+boston_data = load_boston()
```
Comment (harupy) [inline examples/evaluation/evaluate_on_regressor.py:8]: Can we avoid using `load_boston` since it's deprecated?

https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_boston.html

Comment (harupy) [reply examples/evaluation/evaluate_on_regressor.py:8]: Created a PR for this.

File: examples/evaluation/README.md [hunk 16]
```diff
@@ -0,0 +1,27 @@
+### MLflow evaluation Examples
+
+The three simple examples illustrates how you can use the `mlflow.evaluate` API to evaluate a PyFunc model on the
```
Comment (harupy) [inline examples/evaluation/README.md:3]:
```suggestion
The examples in this directory illustrate how you can use the `mlflow.evaluate` API to evaluate a PyFunc model on the
```

File: examples/evaluation/README.md [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+#### Prerequisites
+
+```
+pip install scikit-learn xgboost shap matplotlib
+```
+
+#### How to run this code
+
+Run from the current git directory with Python.
```
Comment (harupy) [inline examples/evaluation/README.md:21]:
```suggestion
Run in this directory with Python.
```

File: examples/evaluation/README.md [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+- Example `evaluate_on_regressor.py` evaluate as scikit-learn `LinearRegression` model on dataset loaded by
+  `sklearn.datasets.fetch_california_housing`
+
+#### Prerequisites
+
+```
+pip install scikit-learn xgboost shap matplotlib
+```
+
+#### How to run this code
```
Comment (harupy) [inline examples/evaluation/README.md:19]:
```suggestion
#### How to run the examples
```

Can we replace `code` because there is no code?

File: examples/evaluation/evaluate_on_regressor.py [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+        targets=y_test,
+        model_type="regressor",
+        dataset_name="california_housing",
+        evaluators="default",
+        feature_names=california_housing_data.feature_names,
+        evaluator_config={"explainability_nsamples": 1000},
+    )
+
+print(f"metrics:\n{result.metrics}")
+print(f"artifacts:\n{result.artifacts}")
```
Comment (harupy) [inline examples/evaluation/evaluate_on_regressor.py:30]: Ran this file on local and these two lines printed out:

```
{'example_count': 6812, 'mean_absolute_error': 0.5295710106684688, 'mean_squared_error': 0.5369686543372452, 'root_mean_squared_error': 0.7327814505957729, 'sum_on_label': 14117.59516999994, 'mean_on_label': 2.0724596550205434, 'r2_score': 0.5970494128783959, 'max_error': 9.907072851035238, 'mean_absolute_percentage_error': 0.3178634328048179}
artifacts:
{}
```

Comment (harupy) [reply examples/evaluation/evaluate_on_regressor.py:30]: It turns out my shap verion is old.

File: examples/evaluation/evaluate_on_binary_classifier.py [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+        model_uri,
+        eval_data,
+        targets="label",
+        model_type="classifier",
+        dataset_name="adult",
+        evaluators=["default"],
+    )
+
+print(f"metrics:\n{result.metrics}")
+print(f"artifacts:\n{result.artifacts}")
```
Comment (harupy) [inline examples/evaluation/evaluate_on_binary_classifier.py:31]: This line's output looks like:

```
{'roc_curve_plot': <mlflow.models.evaluation.artifacts.ImageEvaluationArtifact object at 0x7f5d98244950>, 'precision_recall_curve_plot': <mlflow.models.evaluation.artifacts.ImageEvaluationArtifact object at 0x7f5d982890d0>, 'lift_curve_plot': <mlflow.models.evaluation.artifacts.ImageEvaluationArtifact object at 0x7f5d985bc250>, 'confusion_matrix': <mlflow.models.evaluation.artifacts.ImageEvaluationArtifact object at 0x7f5d9826bc10>, 'shap_beeswarm_plot': <mlflow.models.evaluation.artifacts.ImageEvaluationArtifact object at 0x7f5d98123fd0>, 'shap_summary_plot': <mlflow.models.evaluation.artifacts.ImageEvaluationArtifact object at 0x7f5d2c18fd50>, 'shap_feature_importance_plot': <mlflow.models.evaluation.artifacts.ImageEvaluationArtifact object at 0x7f5d96eb1990>}
```

We might want to implement `__str__` for `EvaluationArtifact` for better string representation (e.g. `<EvaluationArtifact file_name.png>`).

cc @dbczumar

Comment (WeichenXu123) [reply examples/evaluation/evaluate_on_binary_classifier.py:31]: We can do this in follow-up updates.

Comment (WeichenXu123) [reply examples/evaluation/evaluate_on_binary_classifier.py:31]: I add a `__repr__` for evaluation artifact , format is like `ImageEvaluationArtifact(uri='...')`
